lib/ecto/repo.ex

```diff
@@ -818,6 +818,20 @@ defmodule Ecto.Repo do
   This callback is invoked for all query APIs, including the `stream`
   function, but it is not invoked for `insert_all` nor any of the
   schema functions.
+
+  ## Example
+      defmodule Repo do
+        use Ecto.Repo, otp_app: my_app, adapter: Ecto.Adapters.Postgres
+
+        require Ecto.Query
+
+        def prepare_query(op, %{from: %{source: {"posts", _}}} = query, opts) do
+          ammended_query = Ecto.Query.from(s in query, where: s.deleted == true)
+          {ammended_query, opts}
+        end
+
+        def prepare_query(op, query, opts), do: {query, opts}
+      end
```

Comment:
Thanks for the PR but I am not sure if this is a good example. :) The fields from Ecto.Query are available only to adapter implementations and therefore are between public and private APIs.

Reply:
@josevalim how about something like this:

```markdown
## Examples

Let's say you want to filter out records that were "soft-deleted" (have `deleted_at` column set)
from all operations unless an admin is running the query, you can define the callback like this:

    @impl true
    def prepare_query(_operation, query, opts) do
      if opts[:admin] do
        {query, opts}
      else
        query = from(x in query, where: is_nil(x.deleted_at))
        {query, opts}
      end
    end

And then execute the query:

    Repo.all(query)              # only non-deleted records are returned
    Repo.all(query, admin: true) # all records are returned
```

Reply:
I updated it on the branch :)

Reply:
TIL. Cheers! :)